tao_compiler/mlir/disc/transforms/disc_shape_optimization.cc

```diff
@@ -0,0 +1,1474 @@
+
+    rewriter.replaceOp(dimOp, newValue);
+    return success();
+  }
+};
+
+// materialize shape computation IR by resolving tensor.dim + op having
+// interface ShapedTypeOpInterface: before
+//  ```
+//    %arg0 = ... : tensor<?x10xf32>
+//    %0 = mhlo.transpose(%arg0) : tensor<?x?xf32>
+//    %0_d0 = tensor.dim %0, %c0 : tensor<?x?xf32>
+//    %0_d1 = tensor.dim %0, %c1 : tensor<?x?xf32>
+//    %0_new = disc_shape.tie_shape(%0, %0_d0, %0_d1) : tensor<?x?xf32>
+//    use(%0_new)
+//  ```
+// after materializling mhlo.transpose shape interface
+//  ```
+//    %arg0 = ... : tensor<?x10xf32>
+//    %0 = mhlo.transpose(%arg0) {permutaion = [1, 0]} : tensor<?x?xf32>
+//    %0_d0 = tensor.dim %arg0, %c1 : tensor<?x?xf32>
+//    %0_d1 = tensor.dim %arg0, %c0 : tensor<?x?xf32>
+//    %0_new = disc_shape.tie_shape(%0, %0_d0, %0_d1) : tensor<?x?xf32>
+//    use(%0_new)
+//  ```
+LogicalResult materializeShapeComputation(ModuleOp m, FuncOp main) {
+  // Currently we call inline before all disc passes and thus we do not need to
+  // worry about function call ops. Re-visit this once we change the strategy.
+  if (failed(insertTieShapeOnRegion(&main.getBody()))) {
+    return failure();
+  }
+
+  RewritePatternSet patterns(m.getContext());
+  // clang-format off
+  patterns.add<
+      ExpandShapeOfOpPattern,
+      DimOfShapedTypeOpInterface<tensor::DimOp>
+  >(patterns.getContext());
+  // clang-format on
+
+  if (failed(
+          applyPatternsAndFoldGreedily(m->getRegions(), std::move(patterns)))) {
+    return m.emitError() << "fail to materialize shape computation\n";
+  }
+  return success();
+}
+
+/////////////////////////// Stage #1 END //////////////////////////////////////
+
+/////////////////////////// Stage #2 BEGIN ////////////////////////////////////
+
+using PassPipelineRunner =
+    std::function<LogicalResult(OpPassManager&, ModuleOp)>;
+
+// convert:
+//   %1 = disc_shape.tie_shape %0, %d0, %d1, ... : (tensor<?x?xf32>, ...) ->
+//   tensor<?x?xf32> %dim_size = tensor.dim %1[%c0] : tensor<2xindex>
+//   use(%dim_size)
+// to:
+//   use(%d0)
+struct DimOfTieShapeOpCanonicalizationPattern
+    : public OpRewritePattern<tensor::DimOp> {
+  using OpRewritePattern<tensor::DimOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::DimOp op,
+                                PatternRewriter& rewriter) const override {
+    auto tieShapeOp = op.source().getDefiningOp<disc_shape::TieShapeOp>();
+    if (!tieShapeOp) return failure();
+    Optional<int64_t> dimIndex = op.getConstantIndex();
+    if (!dimIndex) return failure();
+    rewriter.replaceOp(op, tieShapeOp->getOperand(1 + *dimIndex));
+    return success();
+  }
+};
+
+// convert:
+//   %1 = disc_shape.tie_shape %0, %d0, %d1, ...
+//         : (tensor<?x?xf32>, ...) -> tensor<?x?xf32>
+//   %2 = tensor.extract %1[...] : ...
+// to:
+//   %2 = tensor.extract %0[...] : ...
+struct ExtractElementOfTieShapeOpCanonicalizationPattern
```

Comment:
We let tensor.extract bypass tie_shape op to make normal canonicalization & cse patterns work. This will be helpful to simplify shape computation IR.

In our current design, tensor.extract op can only be used to shape tensor (e.g. those must be placed on CPU). And all shape tensors are supposed to have static shape (since DISC is a static rank compiler).

Ideally, we should not insert a tie_shape for shape tensors. However, some tensor are known to be "shape tensor" after we insert tie_shape ops. For example, some tensor.extract ops are inserted during we resolve tensor.dim + InferShapedTypeOpInterface.